Adds Subscription API

ikreymer opened the pull request:
Fixes https://github.com/webrecorder/browsertrix/issues/1905

(Different approach than originally suggested, but accomplishes same ends for tracking subscription history in SaaS setup)

Adds a new top-level `/api/subscriptions` endpoint and SubOps handler on the backend.

The subscription api supports `/api/subscriptions/create`, `/api/subscriptions/update` and `/api/subscriptions/cancel`.

- `/api/subscriptions/create` creates a new org with subscription, extending the work done in #1897. The original `/api/orgs/create` is reverted as before, while this endpoint can be used to create a new org with additional options.

- `/api/subscriptions/update` can be used to update subscription data, including status, readOnlyOnCancel and extra details (see below)

- `/api/subscriptions/cancel` can be used to cancel a subscription. By default, this will result in the org also be deleted, unless `readOnlyOnCancel` was previous set.

All subscription events are timestamped and also logged in the MongoDB subscription collection.

All operations can only be accessed via the shared secret auth added in #1897, or with superadmin login.

The Subscription optionally stored on the org in `subscription` is as follows:
```
class Subscription(BaseModel):
    """subscription data"""

    subId: str
    status: str
    planId: str

    futureCancelDate: Optional[datetime] = None
    readOnlyOnCancel: bool = False
```
The `subId` is used to uniquely identify subscriptions, and must be unique if set (via sparse unique index on the Org).
The `planId` field is used to identify an arbitrary plan for the subscription.
`status` and `futureCancelDate` can be used to display status and cancellation info for the user.
`readOnlyOnCancel` indicates where the whole Org should be deleted when a subscription is canceled via `/api/subscriptions/cancel`.

TODOs:
- [x] new POST /subscriptions/create, /subscriptions/update, /subscriptions/cancel API endpoints
- [x] Subscriptions mongo collection storing timestamped /subscription API events
- [x] GET /subscriptions/events API to get subscription events
- [x] Subscription data model 
- [x] Support for setting and handling readOnlyOnCancel on org
- [x] /org/subscription info with external API call to get portalUrl
- [x] subscription in org getter and list views
- [x] subscription API endpoints available only if `billing_enabled` is set in helm chart
- [x] Mark org as readOnly for subscription status `paused_payment_failed`, clears it on status `active`

ikreymer committed:
subscriptions api:
- new subscriptions api which supports /create, /update, /cancel
- move /orgs/create with subscriptions to /subscriptions/create,
keep /orgs/create simple
- store subscription api calls as entries in new subscriptions db, with types 'create', 'update', 'cancel'
- add SubscriptionData type for org.subData, also SubscriptionCreate, SubscriptionUpdate, SubscriptionCancel
- add sparse unique index on org.subData.subId to avoid dupe subs
- tests: move subscription tests to test_org_subs, add test for dupe subscription
- support /subscriptions/update updates SubscriptionData on matching org
- support /subscriptions/cancel delete SubscriptionData on matching org, and deletes the whole org if deleteOnCancel is set to true (the default)

ikreymer committed:
work on tests

ikreymer committed:
fix subId

ikreymer committed:
fix test

ikreymer committed:
remove add_org, simplify dupe check

ikreymer committed:
add /orgs/<id>/read-only-on-cancel option, default to false
tests: test cancelling subscription with and without readOnlyOnCancel set

ikreymer committed:
test fix

ikreymer committed:
more type fixes for org
collection delete: add option to ignore missing collection

ikreymer committed:
additional type fixes

ikreymer committed:
add update test, exclude unset

ikreymer committed:
undo ignore_missing for collections, shouldn't be needed?

ikreymer committed:
add GET /orgs/<id>/subscription endpoint (still needs portalUrl lookup)

ikreymer committed:
lint fixes

ikreymer committed:
keep subs data after cancellation, just mark status as 'canceled'

ikreymer committed:
fix update query

ikreymer committed:
undo change, keep cancel as deleting subData

ikreymer committed:
update test

ikreymer committed:
add portalUrl lookup if external subs app api URL is set (via BTRIX_SUBS_APP_URL)

tw4l commented on backend/btrixcloud/emailsender.py:
```diff
@@ -58,10 +58,10 @@ def _send_encrypted(self, receiver: str, name: str, **kwargs) -> None:
 
         if not self.smtp_server:
-            print("Email: No SMTP Server, not sending", flush=True)
+            # print("Email: No SMTP Server, not sending", flush=True)
```
Seems worth keeping uncommented

ikreymer replied:
Yes, this was mostly for test logs, but probably should be an option?

tw4l replied:
Yeah that makes sense. Similarly, printing out the full email is very useful for dev but just clutter in tests.

ikreymer replied:
Added an option to turn the full email logs on/off:
https://github.com/webrecorder/browsertrix/pull/1914/commits/fb8337c881797449a79fd46e64a0256bcac44bcc

ikreymer committed:
rename SubscriptionData -> Subscription
rename subData -> subscription
remove 'details' dict, just add planId string to Subscription model

ikreymer committed:
Merge branch 'main' into subs-api

ikreymer committed:
add 'subscription' to OrgOut

ikreymer committed:
only enable api if 'billing_enabled' is set to true

ikreymer committed:
fix typo

ikreymer committed:
update SubscriptionOut

ikreymer committed:
add events api endpoint

ikreymer committed:
model grouping

ikreymer committed:
update type

ikreymer committed:
revert firstOrgOwner -> firstOrgAdmin change, just keep as is
always update status, planId, futureCancelDate on sub update for simplicity
add tests for setting and clearing futureCancelDate

ikreymer committed:
makr org as readOnly on certain subscription status changes:
- 'paused_payment_failed' marks org as readOnly / readOnlyReason subscriptionPaused
- 'active' marks org as not readOnly
- update tests

tw4l commented on backend/btrixcloud/invites.py:
```diff
@@ -274,14 +274,17 @@ async def get_invite_out(
 
         if include_first_org_admin:
-            invite_out.firstOrgOwner = True
+            invite_out.firstOrgAdmin = True
```
Does this require a frontend change?

ikreymer replied:
no, actually frontend is checking for `firstOrgAdmin` already, but this was changed to `firstOrgOwner`, so changing it back now to match frontend.

tw4l commented on backend/btrixcloud/subs.py:
```diff
@@ -0,0 +1,236 @@
+        return {"updated": True}
+
+    async def cancel_subscription(self, cancel: SubscriptionCancel) -> dict[str, bool]:
+        """delete subscription data, and if readOnlyOnCancel is true, the entire org"""
```
```suggestion
        """delete subscription data, and unless if readOnlyOnCancel is true, the entire org"""
```

tw4l commented on backend/btrixcloud/subs.py:
```diff
@@ -0,0 +1,236 @@
+
+    async def list_sub_events(self):
+        """list subscription events"""
+        # note: should add sorting/filtering eventually
```
maybe worth implementing this and pagination now? I'm happy to jump in and do that if you'd like :)

tw4l replied:
Done! :)

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -977,16 +988,90 @@ class OrgQuotas(BaseModel):
+
+    subId: str
+    status: str
+    planId: str
```
Is the value of this going to be the name of the plan, or are we keeping a mapping somewhere else of plan ids -> names? If the former, maybe just name it `plan` or `planName`?

ikreymer replied:
Yes, the idea is that it would be an id that could be mapped to a name on the frontend - if it is a name, that makes localization trickier.

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -977,16 +988,90 @@ class OrgQuotas(BaseModel):
 # ============================================================================
-class OrgCreate(BaseModel):
-    """Create a new org"""
+class SubscriptionCreate(BaseModel):
```
Might be a good idea to add the oid to subscription events as well (even if we have to add it after the fact, similar to timestamp) so that we can filter and sort by org - seems like the latter especially may be a common use case.

tw4l replied:
Also done! Added to the subscription events and output models for the list endpoint

tw4l replied:
I know we could get `subId` from the org then filter on that but adding oid directly will save users a step. Can revert if you think it's cleaner the other way.

tw4l committed:
Add pagination, filtering, sorting to subs list endpoint

tw4l committed:
Add oid to subscription events, ability to sort and filter on oid

tw4l committed:
Fix linting

tw4l committed:
Fix linting for real

ikreymer committed:
email: add 'email.log_sent_emails' option in chart value to enable logging sent emails (disable by default)
update no smtp server log message

ikreymer committed:
Update backend/btrixcloud/subs.py

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

SuaYoo commented on backend/btrixcloud/subs.py:
```diff
@@ -0,0 +1,345 @@
+    ) -> dict[str, Any]:
+        """create org for new subscription"""
+        subscription = Subscription(
+            subId=create.subId, status=create.status, planId=create.planId
```
What is the `subId` that's included in the request body? If this is the payment processor subscription ID, perhaps it can be named `providerSubId` to differentiate from the `Subscription` resource ID?

tw4l replied:
It's currently acting as both, the `subId` is essentially the primary key as we have a uniqueness constraint enforced on it.

ikreymer replied:
There's only one id, we use a combination of provider subscription id and its opaque in browsertrix but guaranteed to be unique.

SuaYoo commented on backend/btrixcloud/subs.py:
```diff
@@ -0,0 +1,345 @@
+
+        result = {"added": True, "id": new_org.id}
+
+        if create.firstAdminInviteEmail:
```
`firstAdminInviteEmail` is showing as required in API reference docs, should that be the case?

ikreymer replied:
yes, for `/subscriptions/create`, as it needs to send the email to first admin to be invited to the org in this workflow.

This is separate for regular `/org/create` which is used elsewhere for creating orgs.

ikreymer replied:
heh, this was copied from previous refactor, so can remove this conditional!

SuaYoo commented on backend/btrixcloud/models.py:
```diff
@@ -977,16 +988,113 @@ class OrgQuotas(BaseModel):
+    status: str
+    planId: str
+
+    futureCancelDate: Optional[datetime]
```
Can this be any date time format, or UTC?

ikreymer replied:
It should be datetime UTC, believe it will return an ISO string in UTC.

SuaYoo commented on backend/btrixcloud/subs.py:
```diff
@@ -0,0 +1,345 @@
+
+        return subs, total
+
+    async def get_sub_info(
```
The response body schema isn't displaying properly in API docs, shown as `null`

SuaYoo reviewed (commented):
Some feedback based on API docs!

ikreymer committed:
fix response model for Org

ikreymer committed:
Merge branch 'main' into subs-api

ikreymer committed:
add /orgs/<oid>/billing-portal instead of /orgs/<oid>/subscription since subscription
data is now included in get / list endpoints, update tests

ikreymer committed:
remove unneeded conditional

ikreymer committed:
rename api function

ikreymer committed:
chart: add loading BTRIX_SUBS_APP_URL env var

ikreymer committed:
add api key

ikreymer committed:
more fixes for portalUrl lookup

ikreymer committed:
typo fix

ikreymer committed:
move subscription type field to *Out models

ikreymer committed:
ci: add test for portalUrl using echo server
- add secret with echo server URL before starting
- update k3d-ci and microk8s ci
